File: sdks/python/src/utilities/sanitize_api_call.py
"""
API call event sanitization for the zv1 engine.
Strips auth credentials from headers and URLs before emitting events.
"""
from __future__ import annotations

import logging
import re
from typing import Any
from urllib.parse import parse_qs, urlencode, urlparse, urlunparse

logger = logging.getLogger(__name__)

AUTH_HEADER_PATTERN = re.compile(
    r"^(authorization|x-api-key|api-key|apikey|cookie)$",
    re.IGNORECASE,
)

SENSITIVE_PARAM_PATTERN = re.compile(
    r"^(key|apikey|api_key|token|secret|password|access_token)$",
    re.IGNORECASE,
)


def sanitize_api_call_event(event: dict[str, Any]) -> dict[str, Any]:
    """Sanitize an API call event by stripping auth credentials."""
    sanitized = {**event}

    if "request" in sanitized and sanitized["request"]:
        sanitized["request"] = {**sanitized["request"]}

        # Sanitize headers
        headers = sanitized["request"].get("headers")
        if headers:
            clean_headers = {}
            for key, value in headers.items():
                clean_headers[key] = "[REDACTED]" if AUTH_HEADER_PATTERN.match(key) else value
            sanitized["request"]["headers"] = clean_headers

        # Sanitize URL query params
        url = sanitized["request"].get("url")
        if url:
            try:
                parsed = urlparse(url)
                if parsed.query:
                    params = parse_qs(parsed.query, keep_blank_values=True)
                    redacted_params = {}
                    for key, values in params.items():
                        if SENSITIVE_PARAM_PATTERN.match(key):
                            redacted_params[key] = ["[REDACTED]"]
                        else:
                            redacted_params[key] = values
                    new_query = urlencode(redacted_params, doseq=True)
                    sanitized["request"]["url"] = urlunparse(parsed._replace(query=new_query))
            except Exception:
                pass

    return sanitized


async def emit_api_call_event(
    engine_config: dict[str, Any] | None,
    raw_event: dict[str, Any],
) -> None:
    """Emit an on_api_call event if the callback is configured."""
    if not engine_config or not engine_config.get("on_api_call"):
        return

    event = sanitize_api_call_event(raw_event)

    try:
        callback = engine_config["on_api_call"]
        if callable(callback):
            result = callback(event)
            if hasattr(result, "__await__"):
                await result
    except Exception as e:
        logger.warning(f"Error in on_api_call callback: {e}")

File: sdks/nodejs/src/utilities/sanitizeAPICall.js
/**
 * Sanitize an API call event by stripping auth credentials from headers and URLs.
 *
 * @param {Object} event - Raw API call event
 * @returns {Object} Sanitized event safe for user consumption
 */
export function sanitizeAPICallEvent(event) {
    const sanitized = { ...event };

    if (sanitized.request) {
        sanitized.request = { ...sanitized.request };

        // Sanitize headers - redact auth-related values
        if (sanitized.request.headers) {
            const authPattern = /^(authorization|x-api-key|api-key|apikey|cookie)$/i;
            const cleanHeaders = {};
            for (const [key, value] of Object.entries(sanitized.request.headers)) {
                cleanHeaders[key] = authPattern.test(key) ? '[REDACTED]' : value;
            }
            sanitized.request.headers = cleanHeaders;
        }

        // Sanitize URL query params - redact sensitive values
        if (sanitized.request.url) {
            try {
                const urlObj = new URL(sanitized.request.url);
                const sensitiveParams = /^(key|apikey|api_key|token|secret|password|access_token)$/i;
                for (const key of [...urlObj.searchParams.keys()]) {
                    if (sensitiveParams.test(key)) {
                        urlObj.searchParams.set(key, '[REDACTED]');
                    }
                }
                sanitized.request.url = urlObj.toString();
            } catch (e) {
                // If URL parsing fails, leave as-is
            }
        }
    }

    return sanitized;
}

/**
 * Emit an onAPICall event if the callback is configured.
 * Handles both sync and async callbacks and swallows callback errors.
 *
 * @param {Object} engineConfig - Engine config (may contain onAPICall)
 * @param {Object} rawEvent - Raw unsanitized event data
 */
export async function emitAPICallEvent(engineConfig, rawEvent) {
    if (!engineConfig?.onAPICall) return;

    const event = sanitizeAPICallEvent(rawEvent);

    try {
        const result = engineConfig.onAPICall(event);
        if (result && typeof result.then === 'function') {
            await result;
        }
    } catch (e) {
        // Don't let callback errors break the integration
    }
}
